README.md
<!--
This document is authored using GitHub Flavored Markdown:
https://help.github.com/articles/github-flavored-markdown/
-->

# usn-i18n-nobundle

Java programmer's i18n framework, compile time verifiable, free of
<code>[ResourceBundle](http://docs.oracle.com/javase/8/docs/api/java/util/ResourceBundle.html)</code>.

## Background

Internationalization (i18n) of Java programs typically involves separation data
from code, with message texts being moved to special text resource collections
formatted as
<code>[Properties](http://docs.oracle.com/javase/8/docs/api/java/util/Properties.html)</code>
files and called
[resource bundles](http://docs.oracle.com/javase/8/docs/api/java/util/ResourceBundle.html).
Messages in different languages get into different properties files, and
individual messages are identified by their property keys regardless of
language.

This approach, with messages stored in "almost plain text" format, and messages
in different languages stored in different files, has certain advantages. In
particular, it makes it easy to distribute translation work between translators,
and there is no need for every particular translator to deal with excessive text
data.

Meanwhile, having message texts coupled with code via plain text keys also makes
the whole construction fragile and prone to accumulation of occasional errors.
It is easy to get any particular message "orphaned" due to key being
renamed occasionally, and it is equally easy to obtain code that attempts to use
non-existing messages, with that only being discovered at run time.

Continuous integration and support of internationalized software might benefit
from technology that would allow code and message texts being coupled stronger,
with links being verified at compile time. Programmers might also benefit from
autocompletion aids being available when dealing with text messages.

## Approach

This project offers data structures and control facilities that allow bringing
all message texts back into Java code and handled as ordinary Java variables to
allow compile time verification. All text strings related to the same message in
different languages are organized into one collection-based data structure with
a single identifier. And the programmer never needs to care about language for
any current user. The programmer just needs to make necessary arrangements for
the policy that would govern user-to-language mappings, and the framework will
do the rest.

All this is achieved with the price of text messages being put back into Java
code and looking maybe less like "plain text". And they are all collected
together side-by-side in all languages. There may be different opinions on
whether this is an advantage or not. But for cases when the programmer and the
translator happen to be the same person, the advantage is evident :)

## Prerequisites

The framework currently requires Java 1.7+, as it makes use of
<code>[Locale#toLanguageTag()](https://docs.oracle.com/javase/8/docs/api/java/util/Locale.html#toLanguageTag--)</code>.
This may be changed if really necessary.

## Building

Both [pom.xml](pom.xml) for Maven and [build.xml](build.xml) for Ant are
available. Ant buildfile automatically labels every build with current date.
Maven builds are labeled with project version from the .pom file, but a warning
is issued if this version does not match current date. Artifact/build versions
are currently formatted like "YYYYMMDD" rather than "v1.0", as no distinct
versioning policy has evolved so far.

## Usage – HOW-TO

- download or build the latest 'usn-i18n-nobundle-YYYYMMDD.jar' file and add it
  to your application class path;
- make a decision on the `I18nHandler` subclass you need: whether it should
  depend on some context for obtaining user's locale preferences or should it
  get these preferences from the application directly; hence select an existing
  `I18nHandler` subclass or make a subclass of your own;
- create an instance of the chosen `I18nHandler` subclass in your application;
  no need to care about its further fate; feel free to mark it with
  `@SuppressWarnings ("unused")`;
- create as many of `I18nItem` subclasses' instances as you need for all the
  messages that require internationalization;
- use your messages via corresponding `s(TArg1 arg1)` and similar methods;
- refer to the modest example on the package summary javadocs page and to API
  docs in general when necessary;
- enjoy :)

## Hello application

The 'Hello user' style application below demonstrates several basic concepts:

- shorthand notation for localized messages, taking advantage of static import
  of the `LocalizedMessage#lm()` method;
- instantiation of an `I18nHandler` for the application;
- using an internationalized logger;
- formatting messages with arguments;
- using implicit `toString()` calls instead of explicit `s()` when possible.

```java
package test;

import usn.i18n.nobundle.I15dLogger;
import usn.i18n.nobundle.I15dLoggerFactory;
import usn.i18n.nobundle.I18nHandler;
import usn.i18n.nobundle.I18nHandlerForSingleUser;
import usn.i18n.nobundle.I18nItem0;
import usn.i18n.nobundle.I18nItem1;
import usn.i18n.nobundle.I18nItemAny;

import static usn.i18n.nobundle.LocalizedMessage.lm;

public class HelloUser
  {
    // only I18nItem0 and I18nItemAny message items are currently supported
    // by I15dLogger without confusion...
    public static I18nItem0 LOG_MSG_APP_START = new I18nItem0
        (lm ("en", "Application started."),
         lm ("fr", "Le démarrage de l'application."),
         lm ("ru", "Приложение запущено."));
    public static I18nItemAny LOG_MSG_USER_LOGON = new I18nItemAny
        (lm ("en", "User {0} logged on."),
         lm ("fr", "{0} utilisateur connecté."),
         lm ("ru", "Пользователь {0} вошёл в систему."));
    public static I18nItem1<String> USER_MSG_GREETING = new I18nItem1<String>
      (lm ("en", "Hi {0}!"),
       lm ("fr", "Bonjour {0}!"),
       lm ("ru", "Привет, {0}!"));
    public static I18nItem0 USER_MSG_BYE = new I18nItem0
    (lm ("en", "Nothing left to do, bye!"),
     lm ("fr", "Rien à faire, au revoir!"),
     lm ("ru", "Мы всё сделали, до свидания!"));

    public static void main (String [] args)
      {
        // the 'i18nHandler' variable itself is not needed,
        // we just need to istantiate some 'I18nHandler' subclass,
        // and we do it here by using an instance of 'I18nHandlerForSingleUser'
        // that will use system default locale if no other arrangements are made
        @SuppressWarnings ("unused")
        I18nHandler i18nHandler = new I18nHandlerForSingleUser ();

        // arrange for an i18n-aware logger for our application
        I15dLogger logger = I15dLoggerFactory.getLogger (HelloUser.class);

        // log a simple message without arguments
        logger.info (LOG_MSG_APP_START);

        // learn something about the user to say hello
        String userName = System.getProperty ("user.name");

        // log a message with an argument
        logger.info (LOG_MSG_USER_LOGON, userName);

        // issue a message with an argument to a user
        System.out.println (USER_MSG_GREETING.s (userName));

        // for messages without arguments we can use implicit .toString()
        // instead of explicit .s()
        System.out.println (USER_MSG_BYE);
      } // main
  } // class HelloUser
```

## Licenses

The project is issued and distributed under the following licenses:

- [The BSD 2-Clause License](LICENSE.BSD) - for those who care about legal
  issues seriously;
- [The License in Three Lines](LICENSE.LITL) (LITL) - for those who may like it
  :)

## Distributions

The following distribution types are available:

- plain JAR:
  - usn-i18n-nobundle-YYYYMMDD.jar
- Maven style javadoc and sources JARs:
  - usn-i18n-nobundle-YYYYMMDD-javadoc.jar
  - usn-i18n-nobundle-YYYYMMDD-sources.jar
- full distribution that contains everything:
  - usn-i18n-nobundle-YYYYMMDD-full.jar

<!-- NOTE '../../' below compensate 'blob/master/' that is otherwise added by
          GitHub
          -->
The latest release can be found [here](../../releases/latest).

## TODO

- allow convenient handling of plurals;
- make the Maven artifact for the project available from some public repository.

## See also

- [Look Ma, no ResourceBundle :) Or yet another approach to i18n in Java](http://s-n-ushakov.blogspot.com/2013/02/look-ma-no-resourcebundle-or-yet.html) - initial post at blogger.com
- [Look Ma, no ResourceBundle :) Step 2: Dealing with message arguments](http://s-n-ushakov.blogspot.com/2013/06/look-ma-no-resourcebundle-step-2.html) - the second post at blogger.com


CHANGELOG.txt
2015-01-29 - project re-arranged to be placed in individual repository at GitHub
2015-01-26 - project re-arranged as a 'java' repository branch at GitHub
2015-01-20 - project mavenized
           - project arranged for hosting at GitHub
           - dependency versions updated
           - Ant build adjusted to allow targeting Java 1.7 from under 1.8
           - offline javadocs build implemented for Ant
           - handling of throwables in I15dLogger fixed and made more flexible,
             similar to SLF4J 1.6.0+
           - more I18Item subclasses handled by I15dLogger
           - bug in I18nHandler#findBestLocaleTag() fixed
           - javadoc comments cleaned up
2013-06-18 - demo code fixed in 'package-info.java'
2013-06-17 - first public release with message arguments handling
           - the method for matching locales and languages against I18nItem data
             refactored
           - Ant build file targets restructured
           - javadoc comments for the package updated
2013-05-26 - some methods' visibility changed
           - method I18Item#obtainBestLocaleTag(String[]) renamed to
             getBestLocaleTag
           - public method I18Item#getBestLanguage(String[]) added
           - Ant build file refined
2013-05-20 - refactored to deal with locales rather than languages, with
             compiled MessageFormat option removed
           - I18nItem refactored to have messages map as a member rather than
             a superclass
           - methods taking a Throwable added to I15dLogger
           - 'util' segment removed from exported file names
           - 'changelog.txt' added to the distribution
2013-05-14 - first release with message arguments handling
2013-02-18 - initial public release as concept with I18nItem as all-in-one


src/main/java/usn/i18n/nobundle/package-info.java
/*
 * Copyright (c) 2013, Sergey Ushakov, <s-n-ushakov@yandex.ru>
 * All rights reserved.
 * 
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 * 
 * Redistributions of source code must retain the above copyright notice, this
 * list of conditions and the following disclaimer.
 * 
 * Redistributions in binary form must reproduce the above copyright notice,
 * this list of conditions and the following disclaimer in the documentation
 * and/or other materials provided with the distribution.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 * 
 * See http://opensource.org/licenses/BSD-2-Clause for reference.
 */

/**
 * <p>This package implements an alternative to traditional approach to Java
 * software internationalization and allows localized resources be referenced
 * via individual class member names rather than by arbitrary strings. This
 * allows localized resource references be checked at compile time and also
 * facilitates taking advantage of autocompletion aids provided by IDE's.</p>
 * 
 * @author Sergey Ushakov, s-n-ushakov@yandex.ru
 * @version 2013-06-18
 * 
 * <p>An application using this package would normally:</p>
 * <ul>
 * <li>select one of the two approaches to be used for user locale preferences
 * handling:
 *   <ul>
 *     <li>with user preferences handled and optionally specified by a
 *         {@link usn.i18n.nobundle.I18nHandler I18nHandler} subclass
 *         singleton linked to user preferences in an arbitrary application
 *         specific manner, possibly queried with the application or specified
 *         as part of persisted user preferences in
 *         single-user-per-application-instance cases, etc;
 *     <li>with user preferences handled by a
 *         {@link usn.i18n.nobundle.I18nHandlerInContext
 *         I18nHandlerInContext} subclass singleton capable of detecting user
 *         preferences dynamically from an arbitrary application specific
 *         context; such contexts may be constituted, without limitation, by
 *         {@link javax.servlet.ServletRequest ServletRequest},
 *         {@link javax.servlet.http.HttpSession HttpSession}, etc;
 *     </ul>
 *     in fact there is no tremendous difference between these two approaches
 *     when you start following them... :)</li>
 * <li>subclass either
 *     {@link usn.i18n.nobundle.I18nHandler I18nHandler} or
 *     {@link usn.i18n.nobundle.I18nHandlerInContext I18nHandlerInContext},
 *     providing an application-specific
 *     {@link usn.i18n.nobundle.I18nHandler#getUserLocaleTags()
 *     getUserLocaleTags()} method and optionally an application-specific
 *     {@link usn.i18n.nobundle.I18nHandler#getDefaultLocaleTag()
 *     getDefaultLocaleTag()} method;</li>
 * <li>declare a necessary set of static {@code I18nItem} subclass instances,
 *     one for every message that requires being internationalized; feel free to
 *     have them all in one place or scatter them across your package tree to
 *     your best convenience;</li>
 * <li>use them where necessary via their static member names with
 *     {@link usn.i18n.nobundle.I18nItem0#s() .s()} method being called.</li>
 * </ul>
 * 
 * <p>The following classes are intended for direct use by applications:</p>
 * <ul>
 * <li>handlers:
 *   <ul>
 *     <li>{@link usn.i18n.nobundle.I18nHandler I18nHandler}
 *         &ndash; to be subclassed by multi-user applications in cases when no
 *         explicit user context is available, but it is possible to retrieve
 *         user preferences from the application itself;</li>
 *     <li>{@link usn.i18n.nobundle.I18nHandlerInContext
 *         I18nHandlerInContext} &ndash; to be subclassed by multi-user
 *         applications in cases when some sort of user context is available to
 *         retrieve user preferences from, like
 *         {@link javax.servlet.ServletRequest ServletRequest},
 *         {@link javax.servlet.http.HttpSession HttpSession}, etc;</li>
 *     <li>{@link usn.i18n.nobundle.I18nHandlerForSingleUser
 *         I18nHandlerForSingleUser} &ndash; a ready-to-use handler class
 *         suitable for single-user (e.g. desktop) applications in cases when
 *         it is possible to retrieve user preferences once on application
 *         initialization;</li>
 *     <li>{@link usn.i18n.nobundle.I18nHandlerForServletRequest
 *         I18nHandlerForServletRequest} &ndash; a ready-to-use handler class
 *         suitable for HTTP server applications;</li>
 *     </ul>
 *     you need just to instantiate a handler of your choice, and it will
 *     establish itself as an application-wide singleton by itself;</li>
 * <li>message collection classes to be used as static storage for
 *     multi-locale messages or their formatting patterns; these classes
 *     comprise a two-dimensional matrix according to their properties:
 *   <ul>
 *     <li>depending on the approach to user context handling:
 *       <ul>
 *         <li>classes to be used together with
 *             {@link usn.i18n.nobundle.I18nHandlerInContext
 *             I18nHandlerInContext} subclasses; they have 'InContext' suffix in
 *             their names;</li>
 *         <li>classes to be used together with basic
 *             {@link usn.i18n.nobundle.I18nHandler I18nHandler}
 *             subclasses,
 *             {@link usn.i18n.nobundle.I18nHandlerForSingleUser
 *             I18nHandlerSingleUser} included; they do not have any
 *             context-related suffix in their names;</li>
 *         </ul>
 *       </li>
 *     <li>depending on the number of the arguments they take for formatting:
 *         '0', '1', '2', '3' or 'Any', the latter being the last resort for
 *         large number of arguments, lacking the ability to check argument
 *         types at compile time.</li>
 *     </ul>
 *     </li>
 * <li>internationalized (i15d) wrappers for main <a
 *     href="http://www.slf4j.org/">SLF4J</a> classes for logging:
 *     {@link usn.i18n.nobundle.I15dLogger I15dLogger} and appropriate
 *     factory
 *     {@link usn.i18n.nobundle.I15dLoggerFactory I15dLoggerFactory}.</li>
 * </ul>
 * 
 * <p>An example of "hello world" application:</p>
 * <pre>
 * import usn.i18n.nobundle.I15dLogger;
 * import usn.i18n.nobundle.I15dLoggerFactory;
 * import usn.i18n.nobundle.I18nHandler;
 * import usn.i18n.nobundle.I18nItem0;
 * import usn.i18n.nobundle.I18nItemAny;
 * 
 * import static usn.i18n.nobundle.LocalizedMessage.lm;
 * 
 * public class MyApplication
 *   {
 *     // only 'I18nItemAny' message items are currently supported for logger
 *     static I18nItemAny LOG_MSG_HELLO = new I18nItemAny
 *       (lm ("en", "Hello World!"),
 *        lm ("fr_CA", "Bonjour Monde!"));
 *     public static I18nItem0 USER_MESSAGE_1 = new I18nItem0
 *         (lm ("en", "That's cool!"),
 *          lm ("fr_CA", "C'est le pied!"));
 * 
 *     public static void main (String [] args)
 *       {
 *         // the 'i18nHandler' variable itself is not needed,
 *         // we just need to istantiate some 'I18nHandler' subclass,
 *         // and we do it here subclassing 'I18nHandler' anonymously
 *         {@literal @SuppressWarnings} ("unused")
 *         I18nHandler i18nHandler = new I18nHandler ()
 *           {
 *             {@literal @Override}
 *             protected String getDefaultLocaleTag ()
 *               {
 *                 return "fr_CA";
 *               } // getDefaultLocaleTag
 *             {@literal @Override}
 *             protected String [] getUserLocaleTags ()
 *               {
 *                 // do something application-specific here...
 *               } // getUserLocaleTags
 *           };
 * 
 *         I15dLogger logger =
 *           I15dLoggerFactory.getLogger (MyApplication.class);
 * 
 *         // for messages without arguments we can use .toString() implicitly
 *         // instead of explicit .s()
 *         logger.info (LOG_MSG_HELLO);
 * 
 *         System.out.println (USER_MESSAGE_1.s ());
 *       } // main
 *   } // class MyApplication
 * </pre>
 */
package usn.i18n.nobundle;
